Add PR and release guidelines (#3224)
ethereum/web3.js: CONTRIBUTIONS.md

## Guidelines for Pull Requests and Releases (Web3 1.x)

This document provides some ground rules for contributors (including the maintainer(s) of
the project) about how to make, review and publish changes to 1.x. The most basic requirement is
that **Web3 not break**.

### Pull Requests for substantive changes (e.g. everything except comments and docs)

1. Any PR that introduces a logic change should include tests. (In many cases, the tests will take
more time to write than the actual code).

2. All PRs should sit for 72 hours with the `pleasereview` tag in order to garner feedback.

3. No PR should be merged until it has been reviewed, passes CI, and all reviews' comments are
addressed.

> 4. PRs should have a narrow, well-defined focus and make the smallest set of changes possible to achieve their goal. They should include a clear description in the opening comment.

Reviewer: @cgewecke I think we should use a consistent keyword. I means the PR should be well-defined and `consistent`.

Author: @princesinha19 

To clarify, could you give an example of something that was not consistent? Are you concerned about the word 'narrow' and believe that a PR with several inter-connected goals is consistent and therefore acceptable?

Reviewer: @cgewecke In some PR, people try to create unnecessary typings for example also they do not write code in the same way. Like if we use camelcase for variables then one shouldn't use a snake-case or other cases. Also, One should not use Array<T> for typings if, in the whole module, we have use T[ ], this will violate consistency.

Author: @princesinha19. Ah yes that's a great idea, thanks for that suggestion.

Author: @princesinha19 Added a note about this in 89704e19f986c36173d7dc3989747afce849236b. It says that people should follow the existing conventions and preserve stylistic consistency.

Reviewer: Great Thanks.

Reviewer 2: @princesinha19 Thanks for your input!

approval unless there are issues that would result in:

1. Buggy behaviour.

2. Undue maintenance burden.

3. Pessimisation (i.e. speed reduction / meaningful build-size increases).

4. Feature reduction (i.e. it removes some aspect of functionality that users rely on).

5. Avoidable risk (i.e it's difficult to test or hard to anticipate the implications of, without
being strictly necessary to fix something broken).

### Releases

1. All releases should be proposed in a PR and subject to community review for a minimum of 72 hours (3 days). 

2. Release review periods should be accompanied by a published `rc` version which can be used for
sanity checks / additional testing.

3. During release review, the code is frozen unless new changes are proposed, approved and merged.
Changes reset the release clock and should trigger a new `rc` release.

4. Regular maintainers should manually test the `rc` against a Node project and the published
minified bundle in a browser context. An external reviewer should verify they've done the same.
5. A release PR does have to be approved at least by two known contributors of the web3.js project

### Emergencies

> [THIS SECTION NEEDS WORK after discussing the release procedure more...]

Author: ```suggestion
This topic is under org-wide discussion at https://github.com/ethereum/js-organization/issues/6
```

> ### Guidelines for Pull Requests and Releases (Web3 1.x)

Reviewer 2: ```suggestion
## Guidelines for Pull Requests and Releases (Web3 1.x)
```

This document provides some ground rules for contributors (including the maintainer(s) of
the project) about how to make, review and publish changes to 1.x. The most basic requirement is
that **Web3 not break**.

> ## Pull Requests for substantive changes (e.g. everything except comments and docs)

Reviewer 2: ```suggestion
### Pull Requests for substantive changes (e.g. everything except comments and docs)
```

1. Any PR that introduces a logic change should include tests. (In many cases, the tests will take
more time to write than the actual code).

> 2. All PRs should sit for 48 hours with the `pleasereview` tag in order to garner feedback.

Reviewer 2: ```suggestion
2. All PRs should sit for 72 hours with the `pleasereview` tag in order to garner feedback.
```

3. No PR should be merged until it has been reviewed, passes CI, and all reviews' comments are
addressed.

4. PRs should have a narrow, well-defined focus and make the smallest set of changes possible to achieve their goal. They should include a clear description in the opening comment.

5. Given the choice between a conservative change that mostly works and an adventurous change which
seems better but introduces uncertainty - prefer the conservative change.

> ## Reviews

Reviewer 2: ```suggestion
### Reviews
```

The end-goal of review is to suggest useful improvements to the author. Reviews should finish with
approval unless there are issues that would result in:

1. Buggy behaviour.

2. Undue maintenance burden.

3. Pessimisation (i.e. speed reduction / meaningful build-size increases).

4. Feature reduction (i.e. it removes some aspect of functionality that users rely on).

5. Avoidable risk (i.e it's difficult to test or hard to anticipate the implications of, without
being strictly necessary to fix something broken).

> ## Releases

Reviewer 2: ```suggestion
### Releases
```

> All releases should be proposed in a PR and subject to community review for a minimum of 72 hours (3 days). 

Reviewer 2: ```suggestion
1. All releases should be proposed in a PR and subject to community review for a minimum of 72 hours (3 days). 
```

2. Release review periods should be accompanied by a published `rc` version which can be used for
sanity checks / additional testing.

3. During release review, the code is frozen unless new changes are proposed, approved and merged.
Changes reset the release clock and should trigger a new `rc` release.

4. Regular maintainers should manually test the `rc` against a Node project and the published
minified bundle in a browser context. An external reviewer should verify they've done the same.

> 

Reviewer 2: ```suggestion
5. A release PR does have to be approved at least by two known contributors of the web3.js project

```

Reviewer 2: We should probably also add a point that does define how long an RC release has to be published on NPM until it gets published in the ``latest`` track of npm. One week? @cgewecke

Author: @nivida At the moment the `rc` is tied pretty closely to a code freeze and the review window which is 3 days. Do you want to extend that to a week? Or decouple them?

Reviewer 2: Oh, you thought to release the ``rc`` as soon as the release PR is open and to have the ``rc`` published 3d until we move the changes to the ``latest`` track?
I thought we will keep a release PR open for 3d and after do we release a ``rc`` version which does have to be published for a week until we move it to the ``latest`` track. 

I think from a "never doing a bad release again" perspective does the long process with waiting for a week give us some more safety. But on the other hand, does it extremely increase the time until a fix got released. 🤔

Reviewer 2: >  But on the other hand, does it extremely increase the time until a fix got released.

I think we should probably differentiate between normal releases and ``hot-fix`` releases in this document. In cases of security vulnerabilities or other emergency cases would the goal be I think to have the shortest possible reaction time.

Author: > I thought we will keep a release PR open for 3d and after do we release a rc version which does have to be published for a week until we move it to the latest track.

I think having an `rc` available helps with the review. It lets you install the current state in some project and tests things - you aren't restricted to manually reviewing the code which is often difficult to understand all the implications of.

I'm happy to have a longer review window and a week sounds fine to me. Agree that it delays releases and agree we *need a hot-fix protocol* for releases that fall outside this normal pattern.

Reviewer 2: @cgewecke Great we are on the same page! 💪

Author: @nivida. Lengthened the release review window to 1 week in 89704e19f986c36173d7dc3989747afce849236b and just said that if there are changes during the window, people should have 'enough' time to review a new rc that includes them. 

Just lmk if that's not what you meant here.

> ## Emergencies

Reviewer 2: ```suggestion
### Emergencies
```

> [THIS SECTION NEEDS WORK after discussing the release procedure more...]

Reviewer 2: Let us reference here the issue from the ``js-organization`` repository. (https://github.com/ethereum/js-organization/issues/6)

more time to write than the actual code).

2. All PRs should sit for 48 hours with the `pleasereview` tag in order to garner feedback.

3. No PR should be merged until it has been reviewed, passes CI, and all reviews' comments are
addressed.

4. PRs should have a narrow, well-defined focus and make the smallest set of changes possible to achieve their goal. They should include a clear description in the opening comment.

5. Given the choice between a conservative change that mostly works and an adventurous change which
seems better but introduces uncertainty - prefer the conservative change.

## Reviews

The end-goal of review is to suggest useful improvements to the author. Reviews should finish with
approval unless there are issues that would result in:

1. Buggy behaviour.

2. Undue maintenance burden.

3. Pessimisation (i.e. speed reduction / meaningful build-size increases).

4. Feature reduction (i.e. it removes some aspect of functionality that users rely on).

5. Avoidable risk (i.e it's difficult to test or hard to anticipate the implications of, without
being strictly necessary to fix something broken).

## Releases

> 1. All releases should be proposed in a PR and subject to community review for [SOME_AMOUNT_OF_TIME].

Author: Per meeting with @nivida discussion on Nov 25...
```suggestion
All releases should be proposed in a PR and subject to community review for a minimum of 72 hours (3 days). 
```

### Guidelines for Pull Requests and Releases (Web3 1.x)

This document provides some ground rules for contributors (including the maintainer(s) of
the project) about how to make, review and publish changes to 1.x. The most basic requirement is
that **Web3 not break**.

## Pull Requests for substantive changes (e.g. everything except comments and docs)

1. Any PR that introduces a logic change should include tests. (In many cases, the tests will take
more time to write than the actual code).

> 2. All PRs should sit for 48 hours with the `pleasereview` tag in order to garner feedback.

Reviewer 2: I like that! It's way easier to see with this tag where reviews are required and where they aren't!

3. No PR should be merged until it has been reviewed, passes CI, and all reviews' comments are
addressed.

4. PRs should have a narrow, well-defined focus and make the smallest set of changes possible to achieve
their goal.

5. Given the choice between a conservative change that mostly works and an adventurous change which
seems better but introduces uncertainty - prefer the conservative change.

> 

Reviewer 2: I think we can add here a small sentence about the importance of well-described PR's.

Author: Agree - I added this to the end of number 4 because it seems like those things go together.

## Reviews

The end-goal of review is to suggest useful improvements to the author. Reviews should finish with
approval unless there are issues that would result in:

1. Buggy behaviour.

2. Undue maintenance burden.

3. Pessimisation (i.e. speed reduction / meaningful build-size increases).

4. Feature reduction (i.e. it removes some aspect of functionality that users rely on).

5. Avoidable risk (i.e it's difficult to test or hard to anticipate the implications of, without
being strictly necessary to fix something broken).

## Releases

1. All releases should be proposed in a PR and subject to community review for [SOME_AMOUNT_OF_TIME].

2. Release review periods should be accompanied by a published `rc` version which can be used for
sanity checks / additional testing.

3. During release review, the code is frozen unless new changes are proposed, approved and merged.
Changes reset the release clock and should trigger a new `rc` release.

4. Regular maintainers should manually test the `rc` against a Node project and the published
minified bundle in a browser context. An external reviewer should verify they've done the same.

> ## Emergencies

Reviewer 2: This is something we probably should organize and define EF-JS Team wide? (https://github.com/ethereum/js-organization/)

Author: Yes - I would definitely like more advice about this. Should a bad release be unpublished at npm? If a bad change is introduced, should a release just be rolled back until it has more review? I don't know the best answers to these questions.

Reviewer 2: I've created [here](https://github.com/ethereum/js-organization/issues/6) an issue to discuss this with the required EF projects.
